### PR TITLE
Aux author extras GitHub issue fix 194

### PR DESCRIPTION
**Pull request by rich-hart**
Adding in auxiliary author search when database returns no results on first try.

### CODE REVIEW

**Comment by rich-hart** [comment]
thanks for the comments.  I'll implement them!

File: cnxarchive/views.py [hunk 1]
```diff
@@ -557,6 +561,24 @@ def search(environ, start_response):
     status = '200 OK'
     headers = [('Content-type', 'application/json')]
     start_response(status, headers)
+    
+    #In the case where a search is performed with an authorId as a filter, it is possible for the database to return no results even if the author exists in the database.  Therefore, the database is queried a second time for contact information associated with only the authorIds.  The author information is then used to update the results returned by the first database query.
```
Comment (pumazi) [inline cnxarchive/views.py]: We try to be [PEP8](https://www.python.org/dev/peps/pep-0008/) compliant when we can, that includes a 80char line length limit.

```
> pep8 views.py | grep 565
views.py:565:5: E265 block comment should start with '# '
views.py:565:80: E501 line too long (377 > 79 characters)
views.py:565:378: W291 trailing whitespace
```

Granted, we have more than a few long lines and other PEP8 problems in this file, but nothing over 100 chars. And when only do that when it it would otherwise make the statement look really awful.

**Comment by karenc** [comment]
Instead of putting the information in the branch name, I think it's ok to have a shorter name like "aux_author_extras" and then put the issue you're fixing in the commit message. (Have a look at https://help.github.com/articles/closing-issues-via-commit-messages/)

So usually my commit message would be like this (copying the description from your last pull request):

```
Add aux author data for empty search results

In the case where a search is performed with an authorId as a filter,
it is possible for the database to return no results even if the author
exists in the database. Therefore, the database is queried a second
time for contact information associated with only the authorIds. The
author information is then used to update the results returned by the
first database query.

Fix #194
```

I think if you're going to write a description for the pull request, it's better
to just include it in the commit message.  So we can see everything
just in git without going to github too much.

File: cnxarchive/search.py [hunk 2]
```diff
@@ -654,4 +654,4 @@ def search(query, query_type=DEFAULT_QUERY_TYPE,
             cursor.execute(statement, arguments)
             search_results = cursor.fetchall()
     # Wrap the SQL results.
-    return QueryResults(search_results, query, query_type)
+    return QueryResults(search_results, query, query_type)
```
Comment (karenc) [inline cnxarchive/search.py]: Is it ok to revert the change to this file?  If there are any files with only whitespace changes that are not necessary, do `git checkout cnxarchive/search.py`, so you revert the change before you commit.

File: cnxarchive/sql/get-users-by-ids.sql [hunk 3]
```diff
@@ -0,0 +1,17 @@
+-- ###
+-- ###
+SELECT row_to_json(user_row) 
+FROM   (SELECT username                           AS id, 
+               first_name                         AS firstname, 
+               last_name                          AS surname, 
+               full_name                          AS fullname, 
+               title                              AS title, 
+               (SELECT Array_agg(value)
```
Comment (karenc) [inline cnxarchive/sql/get-users-by-ids.sql]: Could you also change `Array_agg(value)` to uppercase `ARRAY_AGG(value)`?

File: cnxarchive/sql/get-users-by-ids.sql [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+SELECT row_to_json(user_row) 
+FROM   (SELECT username                           AS id, 
+               first_name                         AS firstname, 
+               last_name                          AS surname, 
+               full_name                          AS fullname, 
+               title                              AS title, 
+               (SELECT Array_agg(value) 
+                FROM   contact_infos AS ci 
+                WHERE  ci.user_id = u.id 
+                       AND TYPE = 'EmailAddress') AS emails,
```
Comment (karenc) [inline cnxarchive/sql/get-users-by-ids.sql]: I think `TYPE = 'EmailAddress'` should be lowercase `type = 'EmailAddress'` if `type` is a column name.

File: cnxarchive/database.py [hunk 5]
```diff
@@ -52,7 +52,8 @@ def _read_sql_file(name):
     'get-module-versions': _read_sql_file('get-module-versions'),
     'get-subject-list': _read_sql_file('get-subject-list'),
     'get-featured-links': _read_sql_file('get-featured-links'),
-    }
+    'get-users-by-ids':_read_sql_file('get-users-by-ids'),
+    }
```
Comment (karenc) [inline cnxarchive/database.py]: I'm just going to paste pep8 errors.  Please fix them.  We hope to be pep8 compliant eventually... :)

```
cnxarchive/database.py:55:23: E231 missing whitespace after ':'
cnxarchive/database.py:56:6: W291 trailing whitespace
```

File: cnxarchive/tests/test_views.py [hunk 6]
```diff
@@ -1236,6 +1236,57 @@ def test_search_filter_by_authorID(self):
...
+                    u'fullname': u'OpenStax College',
+                    u'id': u'OpenStaxCollege'
+                }
+            ]
+        ]
+            
+        self.assertEqual( results['results']['auxiliary']['authors'],expected)
+
+
+
```
Comment (karenc) [inline cnxarchive/tests/test_views.py]:
```
cnxarchive/tests/test_views.py:1239:1: W293 blank line contains whitespace
cnxarchive/tests/test_views.py:1241:80: E501 line too long (149 > 79 characters)
cnxarchive/tests/test_views.py:1241:150: W291 trailing whitespace
cnxarchive/tests/test_views.py:1242:1: W293 blank line contains whitespace
cnxarchive/tests/test_views.py:1245:80: E501 line too long (124 > 79 characters)
cnxarchive/tests/test_views.py:1257:19: E222 multiple spaces after operator
cnxarchive/tests/test_views.py:1285:1: W293 blank line contains whitespace
cnxarchive/tests/test_views.py:1286:26: E201 whitespace after '('
cnxarchive/tests/test_views.py:1286:69: E231 missing whitespace after ','
```

File: cnxarchive/views.py [hunk 7]
```diff
@@ -557,6 +561,27 @@ def search(environ, start_response):
...
     start_response(status, headers)
+    
+    #In the case where a search is performed with an authorId as a filter,
+    # it is possible for the database to return no results even if the author exists in the database.
+    #  Therefore, the database is queried a second time for contact information associated with only the authorIds.
+    #  The author information is then used to update the results returned by the first database query. 
+    if len(db_results)<=0 :
+        authors_list =[]
+        keyword_list = query.filters + query.terms + query.sorts
+        for key, value in keyword_list:
```
Comment (karenc) [inline cnxarchive/views.py]: `keyword_list` includes `query.sorts` which is just a single value.  So `keyword_list` looks like this for `cnxarchive.tests.test_views.ViewsTestCase.test_search_unbalanced_quotes`:

```
[('text', 'a phrase'), ('text', 'something else'), ('author', 'first last'), 'pubDate']
```

May be (I'm not sure) it's not necessary to include `query.sorts`?

File: cnxarchive/views.py [hunk 8]
```diff
@@ -557,6 +561,27 @@ def search(environ, start_response):
...
+        if authors_list!=[]:
+            settings = get_settings()
+            with psycopg2.connect(settings[config.CONNECTION_STRING]) as db_connection:
+                with db_connection.cursor() as cursor:
+                    arguments =(authors_list,)
+                    statement=SQL['get-users-by-ids']
+                    cursor.execute(statement,arguments)
+                    authors_db_results=cursor.fetchall()
+            results['results']['auxiliary']['authors'] = authors_db_results
+
```
Comment (karenc) [inline cnxarchive/views.py]:
```
cnxarchive/views.py:564:1: W293 blank line contains whitespace
cnxarchive/views.py:566:80: E501 line too long (101 > 79 characters)
cnxarchive/views.py:567:80: E501 line too long (115 > 79 characters)
cnxarchive/views.py:568:80: E501 line too long (102 > 79 characters)
cnxarchive/views.py:568:103: W291 trailing whitespace
cnxarchive/views.py:569:23: E225 missing whitespace around operator
cnxarchive/views.py:569:26: E203 whitespace before ':'
cnxarchive/views.py:570:23: E225 missing whitespace around operator
cnxarchive/views.py:574:22: E225 missing whitespace around operator
cnxarchive/views.py:576:24: E225 missing whitespace around operator
cnxarchive/views.py:578:80: E501 line too long (87 > 79 characters)
cnxarchive/views.py:580:32: E225 missing whitespace around operator
cnxarchive/views.py:581:30: E225 missing whitespace around operator
cnxarchive/views.py:582:45: E231 missing whitespace after ','
cnxarchive/views.py:583:39: E225 missing whitespace around operator
```

File: development.ini [hunk 9]
```diff
@@ -19,10 +19,11 @@ exports-allowable-types =
     pdf:pdf,application/pdf,PDF,PDF file, for viewing content offline and printing.
     epub:epub,application/epub+zip,EPUB,Electronic book format file, for viewing on mobile devices.
     zip:zip,application/zip,Offline ZIP,An offline HTML copy of the content.  Also includes XML, included media files, and other support files.
-##logging-configuration-filepath = <logging.ini (must be absolute path)>
+##
+#logging-configuration-filepath = /Users/richhart/git/cnx-archive/cnxarchive/default-logging.ini
 
 # OpenStax Accounts database connection info
-accounts.fdw.db-connection-string = dbname=oscaccounts-testing host=localhost port=5432 user=accounts password=accounts
+accounts.fdw.db-connection-string = dbname=accounts host=localhost port=5432 user=accounts password=openstaxaccounts
```
Comment (karenc) [inline development.ini]: I think it's good to change the `dbname` to accounts, but I'd prefer to keep `password` as accounts.  So now `dbname`, `user`, `password` are all accounts.

File: development.ini [hunk 10]
```diff
@@ -3,7 +3,7 @@ use = egg:cnx-archive
 db-connection-string = dbname=cnxarchive user=cnxarchive password=cnxarchive host=localhost port=5432
 # a list of memcache servers separated by whitespace
 # (memcache is disabled if no servers are given)
-memcache-servers =
+memcache-servers = localhost
```
Comment (karenc) [inline development.ini]: I think we shouldn't change the settings for `memcache-servers` and `logging`.

File: cnxarchive/views.py [hunk 11]
```diff
@@ -557,6 +561,27 @@ def search(environ, start_response):
...
+    # it is possible for the database to return no results even if the author exists in the database.
+    #  Therefore, the database is queried a second time for contact information associated with only the authorIds.
+    #  The author information is then used to update the results returned by the first database query. 
+    if len(db_results)<=0 :
+        authors_list =[]
+        keyword_list = query.filters
+        for key, value in keyword_list:
+            if key=='authorID':
+                authors_list.append(value)
+        if authors_list!=[]:
```
Comment (karenc) [inline cnxarchive/views.py]: The code looks simpler if you use `if authors_list:`.

**Comment by karenc** [comment]
Other than the style issues I think the code looks great!
